### PR TITLE
feat : Support for deregistering user defined functions

### LINKED ISSUE #9178: Support deregistering user defined functions
### Is your feature request related to a problem or challenge?

Asked on discord: https://discord.com/channels/885562378132000778/1166447479609376850/1205283780378755156

> Is it possible to deregister a UDF from a SessionContext? I would like to remove all ScalarUDFs with a specific function name from my context, but there's no deregister function.

There are functions to add udf/udaf/udwf to SessionContext but not to remove them

https://github.com/apache/arrow-datafusion/blob/c9e4b7b7c5c2c2e7b3ea01040a6edeafd151410c/datafusion/core/src/execution/context/mod.rs#L816-L827

It is possible to remove tables with `SessionContext::deregister_table`:
https://github.com/apache/arrow-datafusion/blob/c9e4b7b7c5c2c2e7b3ea01040a6edeafd151410c/datafusion/core/src/execution/context/mod.rs#L1039-L1049

### Describe the solution you'd like

Add the corresponding deregister functions to SessionContext

```rust
impl SessionContext {
...
    /// returns the previously registered table/scalar/etc function, if any
    pub fn deregister_udtf(&self, name: &str) -> ,Result<Option<Arc<dyn TableFunctionImpl>)  { ... }
    pub fn deregister_udf(&self) ->  Result<Option<Arc<ScalarUDF>>> {...}
    pub fn deregister_udaf(&self) -> Result<Option<Arc<AggregateUDF>>> {...}
    pub fn deregister_udwf(&self) -> Result<Options<Arc<WindowUDF>>> { ... }
...
```

Ideally, we should also  add the corresponding methods to `FunctionRegistry` and `SessionState`:

```rust
impl FunctionRegistry {
...
    /// returns the previously registered table/scalar/etc function, if any
    /// defaults to "not yet implemented"
    pub fn deregister_udf(&self) ->  Result<Option<Arc<ScalarUDF>>> {...}
    pub fn deregister_udaf(&self) -> Result<Option<Arc<AggregateUDF>>> {...}
    pub fn deregister_udwf(&self) -> Result<Options<Arc<WindowUDF>>> { ... }
...
```

```rust
impl FunctionRegistry for SessionState  {
...
    pub fn deregister_udf(&self) ->  Result<Option<Arc<ScalarUDF>>> {...}
    pub fn deregister_udaf(&self) -> Result<Option<Arc<AggregateUDF>>> {...}
    pub fn deregister_udwf(&self) -> Result<Options<Arc<WindowUDF>>> { ... }
...
```

So the task is to
1.  add the above functions
2.  add basic tests (can adapt existing tests that register functions) that show after `deregister_XXX` the function can not be called anymore

### Describe alternatives you've considered

_No response_

### Additional context

I think this would be a great first project as it:
1. Is well defined
2. Would make a real impact and expose you to the code / tests
3. Would contribute to the overall improvements we are making to the user defined functiosn

### PR DESCRIPTION
**Pull request by mobley-trent**
## Which issue does this PR close?

Closes #9178

### CODE REVIEW

**Comment by alamb** [comment]
This looks good to me @mobley-trent  -- do you think it is ready for review?

**Comment by mobley-trent** [comment]
Hey @alamb I'm done with the final touches.

File: datafusion/core/src/execution/context/mod.rs [hunk 1]
```diff
@@ -848,6 +848,21 @@ impl SessionContext {
...
+    pub fn deregister_udf(&self, name: &str) {
+        self.state.write().deregister_udf(name).ok();
+    }
+
+    /// Deregisters a UDAF within this context.
+    pub fn deregister_udaf(&self, name: &str) {
+        self.state.write().deregister_udaf(name).ok();
+    }
+
+    /// Deregisters a UDTF within this context.
```
Comment (alamb) [inline datafusion/core/src/execution/context/mod.rs:861]:
```suggestion
    /// Deregisters a UDWF within this context.
```

File: datafusion/core/src/execution/context/mod.rs [hunk 2]
```diff
@@ -2247,6 +2292,196 @@ mod tests {
         Ok(())
     }
 
+    #[tokio::test]
+    async fn register_deregister_udf() -> Result<()> {
```
Comment (alamb) [inline datafusion/core/src/execution/context/mod.rs:2296]: Would it be possible to move these tests to https://github.com/apache/arrow-datafusion/blob/main/datafusion/core/tests/user_defined/mod.rs 

Doing so would make them more discoverable as well as likely shorter as you could probably avoid having to define ScalarUDFs

**Review by alamb — APPROVED** [review]
Thank you @mobley-trent  -- this looks great. Nicely documented and nicely tested.  I had some small suggestions, but I also think they could be done as a follow on PR as well

**Review by alamb — APPROVED** [review]
Thank you @mobley-trent 🚀

**Comment by alamb** [comment]
I took the liberty of pushing a fix for `cargo fmt` that was preventing CI from passing

**Comment by alamb** [comment]
Thanks again @mobley-trent

**Pull request merged**